File: scripts/ARMA.R
library(forecast)
library(lubridate)
library(tseries)
library(lmtest)

# Leer y preprocesar los datos
delitos_mensuales = read.csv("delitos_mensuales/delitos_mensuales.csv", encoding = "latin1")
delitos_mensuales$fecha_num = my(delitos_mensuales$fecha_num)

# Crear la serie temporal
ts_delitos <- ts(delitos_mensuales$delitos_cometidos, 
                 start = c(year(min(delitos_mensuales$fecha_num)), month(min(delitos_mensuales$fecha_num))), 
                 frequency = 12)

# Dividir en datos de entrenamiento y prueba (80%-20%)
n <- length(ts_delitos)
train_size <- floor(0.8 * n)
ts_train <- window(ts_delitos, start = c(2012, 1), end = c(2012 + (train_size - 1) %/% 12, (train_size - 1) %% 12 + 1))
ts_test <- window(ts_delitos, start = c(2012 + train_size %/% 12, train_size %% 12 + 1))

########################################################################

ARMA <- Arima(ts_train, order = c(12, 0, 2), include.mean = TRUE,method="ML")
predicciones_prueba_ARMA <- forecast(ARMA, h = length(ts_test))

resultados <- data.frame(
  Fecha = time(ts_test),
  Real = as.numeric(ts_test),
  predicciones_prueba_ARMA = as.numeric(predicciones_prueba_ARMA$mean)
)

ggplot(resultados, aes(x = Fecha)) +
  geom_line(aes(y = Real, color = "blue")) +
  geom_line(aes(y = predicciones_prueba_ARMA, color = "red")) +
  labs(title = "Comparación de modelos ARMA",
       y = "Delitos Cometidos",
       x = "Fecha") +
  theme_minimal()


########################################################################

informacion_arma = function(a,b,columna_entrenamiento){
  mi_ARMA = Arima(columna_entrenamiento, order=c(a,0,b), 
                  include.mean = TRUE, 
                  method="ML")
  pvalores = coeftest(mi_ARMA)[,4]
  akaike = mi_ARMA$aic
  limite_ar = paste0("ar",a)
  limite_ma = paste0("ma",b)
  significativo_ar = limite_ar %in% names(pvalores<0.05)
  significativo_ma = limite_ma %in% names(pvalores<0.05)
  limites_significativos = significativo_ar + significativo_ma
  total_significativos = length(mi_ARMA$coef[pvalores <0.05])
  salida = data.frame("AIC"=akaike,
                      "limites_significativos"=limites_significativos,
                      "total_significativos"=total_significativos)
  rownames(salida)[1] = paste0("ARMA_",a,"_0_",b)
  return(salida)  
}

################################################################################

max_ar <- 12
max_ma <- 12

armas <- list()  # Lista para almacenar los resultados

for (a in 1:max_ar) {
  for (b in 1:max_ma) {
    # Usar tryCatch para manejar errores
    resultado <- tryCatch(
      {
        informacion_arma(a, b, ts_train)  # Intentar ejecutar la función
      },
      error = function(e) {
        # Capturar el error y mostrar en qué paso ocurrió
        cat("Error en a =", a, ", b =", b, ":", conditionMessage(e), "\n")
        return(NULL)  # Devolver NULL para que no agregue nada en caso de error
      }
    )
    # Agregar el resultado a la lista si no es NULL
    if (!is.null(resultado)) {
      armas <- append(armas, list(resultado))
    }
  }
}

resultados_armas = do.call(rbind,armas)

resultados_armas_ordenados = resultados_armas %>%
  arrange(desc(limites_significativos), AIC, total_significativos)


####################################################################


mi_ARMA <- Arima(ts_train, order = c(12, 0, 6), include.mean = TRUE,method="ML")
predicciones_prueba_mi_ARMA <- forecast(mi_ARMA, h = length(ts_test))



resultados <- data.frame(
  Fecha = time(ts_test),
  Real = as.numeric(ts_test),
  predicciones_prueba_mi_ARMA = as.numeric(predicciones_prueba_mi_ARMA$mean)
)


ggplot(resultados, aes(x = Fecha)) +
  geom_line(aes(y = Real, color = "Real")) +
  geom_line(aes(y = predicciones_prueba_mi_ARMA, color = "Predicciones ARMA")) +
  labs(title = "Comparación de modelos ARMA",
       y = "Delitos Cometidos",
       x = "Fecha") +
  scale_color_manual(name = "Modelos", 
                     values = c("Real" = "blue", 
                                "Predicciones ARMA" = "orange")) +
  theme_minimal()

File: scripts/ARMA.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec  6 13:22:48 2024

@author: Usuario
"""

# -*- coding: utf-8 -*-
"""
Created on Thu Dec  5 09:21:13 2024

@author: Usuario
"""

import pandas as pd
import numpy as np
from plotnine import *
from mizani.formatters import date_format
from mizani.breaks import date_breaks

import os

#### Para hacer la prueba DF
import statsmodels.tsa.stattools as sts

#### Para analizar la estacionalidad
from statsmodels.tsa.seasonal import seasonal_decompose

#### Para analizar autocorrelación
import statsmodels.graphics.tsaplots as sgt  
from statsmodels.tsa.stattools import acf, pacf

#### Para medias móviles
from statsmodels.tsa.holtwinters import SimpleExpSmoothing, ExponentialSmoothing


#### Para modelos autorregresivos: AR
from statsmodels.tsa.ar_model import AutoReg, ar_select_order

#### Para modelos de medias móviles: MA
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import arma_order_select_ic


import warnings 
warnings.filterwarnings("ignore")


#%%

# Establecer el directorio de trabajo (ajustar según tu sistema)
os.chdir("C:\\Users\\Usuario\\Documents\\scidata\\24_st\\data")

# Leer los archivos CSV
delitos_mensuales = pd.read_csv("delitos_mensuales\\delitos_mensuales.csv",encoding='latin1')
tipo_cambio = pd.read_csv("tipo_de_cambio\\tipo_cambio.csv",encoding='latin1')
climas_nyc = pd.read_csv("climas_nyc\\climas_procesado.csv")
ruidos = pd.read_csv("ruidos\\ruidos.csv")

#%%

############ Preprocesamiento

delitos_mensuales["fecha_num"] = pd.to_datetime(delitos_mensuales["fecha_num"], format='%m/%Y')
delitos_mensuales.index = delitos_mensuales["fecha_num"]

tipo_cambio["Fecha"] = pd.to_datetime(tipo_cambio["Fecha"], format='%d/%m/%Y')
tipo_cambio.index = tipo_cambio["Fecha"]


climas_nyc["datetime"] = pd.to_datetime(climas_nyc['datetime'], format='%d/%m/%Y %H:%M')
climas_nyc.index = climas_nyc["datetime"]

ruidos["fecha"] = pd.to_datetime(ruidos['fecha'], format='%d/%m/%Y')
ruidos.index = ruidos["fecha"]


# Establecemos la frecuencia de la serie de tiempo. En este caso es MS: month start
delitos_mensuales = delitos_mensuales.asfreq("MS")

tipo_cambio = tipo_cambio.asfreq("b")  #bussines 
tipo_cambio['Cambio'] = tipo_cambio['Cambio'].fillna(method='ffill')

climas_nyc = climas_nyc.asfreq("h")

ruidos = ruidos.asfreq("d")

#%%
############### Entrenamiento y prueba

## len(tabla)
n = len(delitos_mensuales)
total_datos_entrenamiento = int(0.8*n)

## tabla.iloc
entrenamiento = delitos_mensuales.iloc[:total_datos_entrenamiento]
prueba = delitos_mensuales.iloc[total_datos_entrenamiento:]

#%%

resultados = pd.DataFrame({"Real": prueba["delitos_cometidos"].values},
                          index = prueba.index)

#%%

# Entrenar un modelo ARMA(1,0,1) manualmente
ARMA_1_0_1 = ARIMA(entrenamiento["delitos_cometidos"], order=(1, 0, 1))  # (p=0, d=0, q=1)
ARMA_1_0_1_ajustado = ARMA_1_0_1.fit()
print(ARMA_1_0_1_ajustado.summary())

# Predicciones sobre los datos de prueba
start = len(entrenamiento)
end = len(entrenamiento) + len(prueba) - 1
predicciones_prueba_ARMA_1_0_1 = ARMA_1_0_1_ajustado.predict(start=start, end=end)

resultados["Prediccion_prueba_ARMA_1_0_1"] = predicciones_prueba_ARMA_1_0_1

resultados.plot(figsize=(12,6))

#%%


# Entrenar el modelo MA con el orden óptimo seleccionado
def informacion_arma(a,b,columna_entrenamiento):
    mi_ARMA = ARIMA(columna_entrenamiento, order=(a,0,b))
    mi_ARMA_ajustado = mi_ARMA.fit()
    akaike = mi_ARMA_ajustado.aic
    limite_ar = "ar.L" + str(a)
    limite_ma = "ma.L" + str(b)
    significativo_ar = limite_ar in mi_ARMA_ajustado.params[mi_ARMA_ajustado.pvalues <0.05].index
    significativo_ma = limite_ma in mi_ARMA_ajustado.params[mi_ARMA_ajustado.pvalues <0.05].index
    limites_significativos = significativo_ar + significativo_ma
    total_significativos = len(mi_ARMA_ajustado.params[mi_ARMA_ajustado.pvalues <0.05])-1
    return pd.DataFrame({"AIC":akaike,
                         "limites_significativos":limites_significativos,
                         "total_significativos":total_significativos},index=[f"ARMA_{a}_0_{b}"])

#%%
max_ar = 6
max_ma = 6

armas = [informacion_arma(a,b,entrenamiento["delitos_cometidos"]) for a in range(1,max_ar+1) for b in range(1,max_ma+1)]

resultados_armas = pd.concat(armas)

resultados_armas_ordenados = resultados_armas.sort_values(
    by=['limites_significativos', 'AIC', 'total_significativos'], 
    ascending=[False, True, True]
)

resultados_armas_ordenados

#%%

mi_ARMA = ARIMA(entrenamiento["delitos_cometidos"], order=(4,0,6))
mi_ARMA_ajustado = mi_ARMA.fit()
print(mi_ARMA_ajustado.summary())

start = len(entrenamiento)
end = len(entrenamiento) + len(prueba) - 1
predicciones_prueba_mi_ARMA = mi_ARMA_ajustado.predict(start=start, end=end)


resultados["Prediccion_prueba_mi_ARMA"] = predicciones_prueba_mi_ARMA
resultados.plot(figsize=(12,6))

